pjlib-util/src/pjlib-util/resolver.c

```
@@ -559,12 +559,8 @@ PJ_DEF(pj_status_t) pj_dns_resolver_set_ns( pj_dns_resolver *resolver,
         if (status != PJ_SUCCESS)
             status = pj_sockaddr_init(pj_AF_INET6(), &ns->addr, &servers[i], 
                                       (pj_uint16_t)(ports ? ports[i] : PORT));
-        if (status != PJ_SUCCESS) {
-            pj_grp_lock_release(resolver->grp_lock);
-            return PJLIB_UTIL_EDNSINNSADDR;
-        }
 
-        ns->state = STATE_ACTIVE;
+        ns->state = (status == PJ_SUCCESS) ? STATE_ACTIVE : STATE_BAD;
         ns->state_expiry = now;
         ns->rt_delay.sec = 10;
     }
```

Actually [the docs](https://docs.pjsip.org/en/latest/api/generated/pjlib_util/group/group__PJ__DNS__RESOLVER.html#_CPPv422pj_dns_resolver_set_nsP15pj_dns_resolverjA_K8pj_str_tA_K11pj_uint16_t) specifies that:
> If hostname is specified, the hostname must be resolvable with pj_gethostbyname(). 

so the behaviour of this function should already be correct.

Also note that if the failure is ignored, the specified nameserver will still be set but has no IP address (due to invalid IP address or unresolvable). Moreover, when all specified nameservers are invalid or not-resolvable, the function seems to still returning success.

So if the objective is to avoid library initialization failure (due to bad nameserver settings), I think the modification should be inside `pjsua_init()`, i.e: validate & filter the nameservers, before calling `pj_dns_resolver_set_ns()`?

---

@nanangizz

According to the same documentation, the second nameserver (then the third and fourth) should be tried, but this will never happen if the company-specific nameserver1 is not whitelisted/accessible due to the firewall/router settings. As a result, we cannot fallback to the Google public DNS 8.8.8.8 in some WiFi networks. 

I agree that the pj_dns_resolver_set_ns() should return success only if at least one nameserver is marked as ACTIVE, so pjsua_init() will throw an error only if all nameservers are BAD.

---

Well, I agree with you (about IP addresses resolving on the client side) if we're talking about the IP phones and desktop clients, but not mobile apps, which can change the network every minute, so I was assuming the 4 nameservers array was introduced to have the ability to switch to the any active one at any time.

Ok, in general yes, this PR needs to be rejected.